Prototype of NF4 RNAseq mini-course (#526)
nextflow-io/training: docs/nf4_science/rnaseq/03_multi-sample.md

include { HISAT2_ALIGN } from './modules/hisat2_align.nf'
include { MULTIQC } from './modules/multiqc.nf'
```

### 2.3. Add a `report_id` parameter and give it a sensible default

```groovy title="rnaseq.nf" linenums="9"
/*
 * Pipeline parameters
 */
params.hisat2_index_zip = "data/genome_index.tar.gz"
params.report_id = "all_single-end"
```

### 2.4. Call the process on the outputs of the previous steps

We need to give the `MULTIQC` process all the QC-related outputs from previous steps.

For that, we're going to use the `.mix()` operator, which aggregates multiple channels into a single one.

If we had four processes called A, B, C and D with a simple `.out` channel each, the syntax would look like this: `A.out.mix( B.out, C.out, D.out )`. As you can see, you apply it to the first of the channels you want to combine (doesn't matter which) and just add all the others, separated by commas, in the parenthesis that follows.

In the case of our workflow, we have the following outputs to aggregate:

- `FASTQC.out.zip`
- `FASTQC.out.html`
- `TRIM_GALORE.out.trimming_reports`
- `TRIM_GALORE.out.fastqc_reports`
- `HISAT2_ALIGN.out.log`

So the syntax example becomes:

> ```groovy title="Applying .mix() in the MULTIQC call"

Reviewer: ```suggestion
```groovy title="Applying .mix() for the MULTIQC call"
```

>         FASTQC.out.zip.mix(

Reviewer: ```suggestion
        Channel.empty.mix(
            FASTQC.out.zip,
```

Author: TIL

Author: Same as below, going to keep that for the next pass of improvements. Great idea, just can't deal with domino updates right now.

        FASTQC.out.html,
        TRIM_GALORE.out.trimming_reports,
        TRIM_GALORE.out.fastqc_reports,
        HISAT2_ALIGN.out.log
        )
```

> To this, we need to append the `collect()` operator in order to pull the reports for all the samples into a single call to `MULTIQC`, and also give it the `report_id` parameter:

Reviewer: ```suggestion
Before we can use the files in the channel we need to use the `collect()` operator in order to pull the reports for all the samples into a single call to `MULTIQC`.
We also need to give it the `report_id` parameter:
```

```groovy title="The completed MULTIQC call" linenums="33"

>     // Comprehensive QC report generation
>     MULTIQC(
>         FASTQC.out.zip.mix(
>         FASTQC.out.html,
>         TRIM_GALORE.out.trimming_reports,
>         TRIM_GALORE.out.fastqc_reports,
>         HISAT2_ALIGN.out.log
>         ).collect(),
>         params.report_id
>     )

Reviewer: With beginners I might suggest splitting these calls for better clarity.
```suggestion
    // Comprehensive QC report generation
    multiqc_files_ch = Channel.empty.mix(
        FASTQC.out.zip,
        FASTQC.out.html,
        TRIM_GALORE.out.trimming_reports,
        TRIM_GALORE.out.fastqc_reports,
        HISAT2_ALIGN.out.log
        )
    multiqc_files_list = multiqc_files_ch.collect(),
    MULTIQC(
        multiqc_files_list,
        params.report_id
    )
```

Author: Ohh that is clearer, yes, thanks

Author: I'm going to keep this as an improvement for the next round because if I make this change there's a bunch of other files that need to be updated, and I don't quite have the runway for that right now. But it's a really good idea for explaining this bit more clearly.
